stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+// semaphore standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SEMAPHORE_
+#define _SEMAPHORE_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <semaphore> is not supported when compiling with /clr:pure.
+#endif // _M_CEE_PURE
+
+#if !_HAS_CXX20
+#pragma message("The contents of <semaphore> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+
+#include <atomic>
+#include <chrono>
+#include <limits.h>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+template <class _Rep, class _Period>
+_NODISCARD unsigned long long _Semaphore_deadline(const chrono::duration<_Rep, _Period>& _Rel_time) {
+    return __std_atomic_wait_get_deadline(
+        chrono::duration_cast<chrono::duration<unsigned long long, milli>>(_Rel_time).count());
+}
+
+template <class _Clock, class _Duration>
+_NODISCARD unsigned long _Semaphore_remaining_timeout(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+    auto _Now = _Clock::now();
+    if (_Now >= _Abs_time) {
+        return 0;
+    }
+
+    auto _Rel_time = _Abs_time - _Now;
+    static constexpr chrono::hours _Ten_days{24 * 10};

```

Comment:
I don't think we should use this 10 day workaround here; elsewhere that was a ninja-patch to avoid some ABI problems. Here we should just try to put the thing in an unsigned, and if that doesn't fit, clamp to `UINT_MAX-1`; for example:

```
auto _Rel_time = ceil<milliseconds>(_Abs_time - _Now).count();
if (_Rel_time >= ULONG_MAX) {
    _Rel_time = ULONG_MAX - 1;
}

return static_cast<unsigned long>(_Rel_time);
```

Reply:
I'm afraid as many Windows APIs don't take `INFINITY-1` as a valid timeout, and given atomic wait fallback may change, I don't want to run into that.

Reply:
OK, then use -2. The point is we should do the clamping entirely in the unsigned long domain rather than bouncing around through the chrono guts.

(But in this case INFINITY-1 should be accepted as it is immediately turned into an NT timestamp under the covers)

Reply:
I see, I thought `chrono` use is desirable.
I observed accepting 10 days, but not `INFINITY-1` in `SendMesssageTimeout`, so I'm paranoid about others.

Reply:
Chrono use is nice but here it could introduce an extra multiply in the `>=` comparing `_Rel_time` and `_Ten_days`. We know we will want the result in milliseconds to pass down anyway, so going to milliseconds and doing all the comparisons that way makes sure we have done as few conversions as we can.

I suppose it isn't a huge deal. I probably overreacted here just seeing the `_Ten_days` thing because that's one of the most horrible ABI workarounds we have ever attempted, so needing it here was ... scary. But the reason it's terrible in the other ABI workaround places doesn't really apply here since we aren't trigging the many repeated clock calls, crazy overflow semantics, and similar.

I wouldn't block over this but still have a mild preference for doing this clamping entirely in the milliseconds domain.